File: recipes/hostui/greeting-hostui-qt/src/DaemonProcess.cpp
#include "DaemonProcess.h"

#include <QCoreApplication>
#include <QDir>
#include <QFile>
#include <QFileInfo>

#include "holons/holons.hpp"

namespace {
constexpr auto kPackagedHolonSlug = "greeting-daemon";
constexpr auto kPackagedHolonUUID = "6492d55a-55b8-4ecb-a406-2a2a401f7c01";
constexpr auto kPackagedFamilyName = "daemon";
constexpr auto kDevDaemonBinaryName = "gudule-daemon-greeting-go";
} // namespace

DaemonProcess::DaemonProcess(const QString &binaryName, QObject *parent)
    : QObject(parent), binaryName_(binaryName) {
}

bool DaemonProcess::start() {
  if (channel_) {
    return true;
  }
  lastError_.clear();
  grpcTarget_.clear();

  const QString binaryPath = resolveBinaryPath();
  if (binaryPath.isEmpty()) {
    lastError_ = QStringLiteral("Daemon binary not found: %1").arg(binaryName_);
    return false;
  }
  binaryPath_ = binaryPath;

  const QString overrideTarget =
      qEnvironmentVariable("GUDULE_DAEMON_TARGET").trimmed();
  if (!overrideTarget.isEmpty()) {
    try {
      channel_ = holons::connect(overrideTarget.toStdString());
      grpcTarget_ = QString::fromStdString(holons::channel_target(channel_));
      if (grpcTarget_.isEmpty()) {
        grpcTarget_ = overrideTarget;
      }
      return true;
    } catch (const std::exception &ex) {
      lastError_ = QString::fromUtf8(ex.what());
      if (channel_) {
        try {
          holons::disconnect(channel_);
        } catch (const std::exception &) {
        }
        channel_.reset();
      }
      grpcTarget_.clear();
      return false;
    }
  }

  auto stageRoot = std::make_unique<QTemporaryDir>(
      QDir::temp().filePath(QStringLiteral("greeting-daemon-stage-XXXXXX")));
  if (!stageRoot->isValid()) {
    lastError_ = QStringLiteral("Failed to create temporary holon root.");
    return false;
  }

  const QDir root(stageRoot->path());
  const QString holonDir =
      root.filePath(QStringLiteral("holons/%1").arg(QString::fromUtf8(kPackagedHolonSlug)));
  if (!QDir().mkpath(holonDir)) {
    lastError_ = QStringLiteral("Failed to create staged holon directory: %1").arg(holonDir);
    return false;
  }

  QFile manifest(QDir(holonDir).filePath(QStringLiteral("holon.yaml")));
  if (!manifest.open(QIODevice::WriteOnly | QIODevice::Text | QIODevice::Truncate)) {
    lastError_ = QStringLiteral("Failed to write staged holon manifest: %1")
                     .arg(manifest.errorString());
    return false;
  }
  manifest.write(buildManifest(binaryPath).toUtf8());
  manifest.close();

  const QString previousDirectory = QDir::currentPath();
  if (!QDir::setCurrent(stageRoot->path())) {
    lastError_ = QStringLiteral("Failed to enter staged holon root: %1").arg(stageRoot->path());
    return false;
  }

  try {
    channel_ = holons::connect(kPackagedHolonSlug);
    grpcTarget_ = QString::fromStdString(holons::channel_target(channel_));
    if (grpcTarget_.isEmpty()) {
      throw std::runtime_error("cpp-holons did not expose the daemon target");
    }
    stageRoot_ = std::move(stageRoot);
  } catch (const std::exception &ex) {
    lastError_ = QString::fromUtf8(ex.what());
    if (channel_) {
      try {
        holons::disconnect(channel_);
      } catch (const std::exception &) {
      }
      channel_.reset();
    }
    grpcTarget_.clear();
  }

  QDir::setCurrent(previousDirectory);
  return static_cast<bool>(channel_);
}

void DaemonProcess::stop() {
  if (channel_) {
    try {
      holons::disconnect(channel_);
    } catch (const std::exception &ex) {
      lastError_ = QString::fromUtf8(ex.what());
    }
    channel_.reset();
  }

  stageRoot_.reset();
  grpcTarget_.clear();
}

std::shared_ptr<grpc::Channel> DaemonProcess::channel() const { return channel_; }

QString DaemonProcess::target() const {
  const QString overrideTarget =
      qEnvironmentVariable("GUDULE_DAEMON_TARGET").trimmed();
  if (!overrideTarget.isEmpty()) {
    return overrideTarget;
  }
  return QString::fromUtf8(kPackagedHolonSlug);
}

QString DaemonProcess::grpcTarget() const { return grpcTarget_; }

QString DaemonProcess::binaryPath() const { return binaryPath_; }

QString DaemonProcess::lastError() const { return lastError_; }

QString DaemonProcess::resolveBinaryPath() const {
  const QDir currentDir(QDir::currentPath());
  const QDir appDir(QCoreApplication::applicationDirPath());
  const QStringList candidates = {
      appDir.filePath(QStringLiteral("daemon/%1").arg(binaryName_)),
      currentDir.filePath(binaryName_),
      currentDir.filePath(QStringLiteral("../../daemons/greeting-daemon-go/.op/build/bin/%1")
                              .arg(QString::fromUtf8(kDevDaemonBinaryName))),
      currentDir.filePath(QStringLiteral("../../daemons/greeting-daemon-go/%1")
                              .arg(QString::fromUtf8(kDevDaemonBinaryName))),
      appDir.filePath(QStringLiteral("../Resources/%1").arg(binaryName_)),
  };

  for (const QString &candidate : candidates) {
    if (QFileInfo::exists(candidate)) {
      return QDir::cleanPath(candidate);
    }
  }

  return {};
}

QString DaemonProcess::buildManifest(const QString &binaryPath) const {
  QString escapedPath = binaryPath;
  escapedPath.replace(QStringLiteral("\\"), QStringLiteral("\\\\"));
  escapedPath.replace(QStringLiteral("\""), QStringLiteral("\\\""));

  return QStringLiteral(
             "schema: holon/v0\n"
             "uuid: \"%1\"\n"
             "given_name: greeting\n"
             "family_name: \"%2\"\n"
             "motto: Packaged greeting daemon fallback.\n"
             "composer: Codex\n"
             "clade: deterministic/pure\n"
             "status: draft\n"
             "born: \"2026-03-11\"\n"
             "generated_by: codex\n"
             "kind: native\n"
             "build:\n"
             "  runner: recipe\n"
             "artifacts:\n"
             "  binary: \"%3\"\n")
      .arg(QString::fromUtf8(kPackagedHolonUUID), QString::fromUtf8(kPackagedFamilyName), escapedPath);
}

File: recipes/hostui/greeting-hostui-dotnet/Services/DaemonProcess.cs
using System.Diagnostics;
using System.Runtime.InteropServices;
using Grpc.Net.Client;
using Holons;

namespace Greeting.Godotnet.Services;

public sealed class DaemonProcess : IAsyncDisposable
{
    private const string PackagedHolonSlug = "greeting-daemon";
    private const string PackagedBinaryName = "gudule-greeting-daemon";
    private const string DevBinaryName = "gudule-daemon-greeting-go";
    private GrpcChannel? _channel;
    private GreetingClient? _client;
    private string? _stageRoot;

    public string HolonSlug => DaemonTargetOverride() ?? PackagedHolonSlug;

    public GreetingClient Client =>
        _client ?? throw new InvalidOperationException("Daemon is not connected.");

    public async Task StartAsync(CancellationToken cancellationToken = default)
    {
        if (_client is not null)
        {
            return;
        }

        AppContext.SetSwitch("System.Net.Http.SocketsHttpHandler.Http2UnencryptedSupport", true);

        var overrideTarget = DaemonTargetOverride();
        if (overrideTarget is not null)
        {
            _channel = Connect.ConnectTarget(overrideTarget);
            _client = new GreetingClient(_channel);
            return;
        }

        var binaryPath = await ResolveBinaryPathAsync(cancellationToken);
        var stageRoot = await StageHolonRootAsync(binaryPath, cancellationToken);
        _stageRoot = stageRoot;

        var previousDirectory = Directory.GetCurrentDirectory();
        try
        {
            cancellationToken.ThrowIfCancellationRequested();
            Directory.SetCurrentDirectory(stageRoot);
            _channel = Connect.ConnectTarget(PackagedHolonSlug);
            _client = new GreetingClient(_channel);
        }
        catch
        {
            if (_channel is not null)
            {
                Connect.Disconnect(_channel);
                _channel = null;
            }

            _client = null;
            _stageRoot = null;
            await DeleteStageRootAsync(stageRoot);
            throw;
        }
        finally
        {
            Directory.SetCurrentDirectory(previousDirectory);
        }
    }

    public async ValueTask DisposeAsync()
    {
        _client = null;
        var root = _stageRoot;
        _stageRoot = null;

        try
        {
            if (_channel is not null)
            {
                Connect.Disconnect(_channel);
                _channel = null;
            }
        }
        finally
        {
            if (root is not null)
            {
                await DeleteStageRootAsync(root);
            }
        }
    }

    private async Task<string> ResolveBinaryPathAsync(CancellationToken cancellationToken)
    {
        var packagedPath = await TryResolvePackagedBinaryAsync(cancellationToken);
        if (packagedPath is not null)
        {
            return packagedPath;
        }

        var candidates = new[]
        {
            Path.GetFullPath(Path.Combine(Directory.GetCurrentDirectory(), PackagedBinaryNameWithHostSuffix())),
            Path.GetFullPath(Path.Combine(Directory.GetCurrentDirectory(), "..", "..", "daemons", "greeting-daemon-go", ".op", "build", "bin", DevBinaryNameWithHostSuffix())),
            Path.GetFullPath(Path.Combine(Directory.GetCurrentDirectory(), "..", "..", "daemons", "greeting-daemon-go", DevBinaryNameWithHostSuffix())),
            Path.GetFullPath(Path.Combine(AppContext.BaseDirectory, PackagedBinaryNameWithHostSuffix())),
        };
        foreach (var candidate in candidates)
        {
            if (File.Exists(candidate))
            {
                return candidate;
            }
        }

        throw new FileNotFoundException($"Unable to locate {PackagedBinaryNameWithHostSuffix()} in the app bundle or source tree.");
    }

    private async Task<string?> TryResolvePackagedBinaryAsync(CancellationToken cancellationToken)
    {
        var binaryName = PackagedBinaryNameWithHostSuffix();
        var cachePath = Path.Combine(FileSystem.Current.CacheDirectory, binaryName);
        if (File.Exists(cachePath))
        {
            return cachePath;
        }

        try
        {
            await using var packageStream = await FileSystem.Current.OpenAppPackageFileAsync(binaryName);
            await using var output = File.Create(cachePath);
            await packageStream.CopyToAsync(output, cancellationToken);
            await output.FlushAsync(cancellationToken);
            await MakeExecutableAsync(cachePath, cancellationToken);
            return cachePath;
        }
        catch (FileNotFoundException)
        {
            return null;
        }
    }

    private static async Task MakeExecutableAsync(string path, CancellationToken cancellationToken)
    {
        if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
        {
            return;
        }

        var chmod = new ProcessStartInfo
        {
            FileName = "/bin/chmod",
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            UseShellExecute = false,
        };
        chmod.ArgumentList.Add("+x");
        chmod.ArgumentList.Add(path);

        using var process = Process.Start(chmod) ?? throw new InvalidOperationException("Unable to invoke chmod.");
        await process.WaitForExitAsync(cancellationToken);
    }

    private async Task<string> StageHolonRootAsync(string binaryPath, CancellationToken cancellationToken)
    {
        var root = Path.Combine(
            FileSystem.Current.CacheDirectory,
            "holons-runtime",
            PackagedHolonSlug);
        var holonDir = Path.Combine(root, "holons", PackagedHolonSlug);
        Directory.CreateDirectory(holonDir);
        await File.WriteAllTextAsync(
            Path.Combine(holonDir, "holon.yaml"),
            BuildManifest(binaryPath),
            cancellationToken);
        return root;
    }

    private string BuildManifest(string binaryPath)
    {
        var escapedPath = binaryPath.Replace("\\", "\\\\", StringComparison.Ordinal)
            .Replace("\"", "\\\"", StringComparison.Ordinal);
        return $"""
schema: holon/v0
uuid: "6492d55a-55b8-4ecb-a406-2a2a401f7c01"
given_name: greeting
family_name: "daemon"
motto: Packaged greeting daemon fallback.
composer: Codex
clade: deterministic/pure
status: draft
born: "2026-03-11"
generated_by: codex
kind: native
build:
  runner: recipe
artifacts:
  binary: "{escapedPath}"
""" + Environment.NewLine;
    }

    private static string? DaemonTargetOverride()
    {
        var value = Environment.GetEnvironmentVariable("GUDULE_DAEMON_TARGET");
        return string.IsNullOrWhiteSpace(value) ? null : value.Trim();
    }

    private static string PackagedBinaryNameWithHostSuffix() =>
        OperatingSystem.IsWindows() ? $"{PackagedBinaryName}.exe" : PackagedBinaryName;

    private static string DevBinaryNameWithHostSuffix() =>
        OperatingSystem.IsWindows() ? $"{DevBinaryName}.exe" : DevBinaryName;

    private static Task DeleteStageRootAsync(string root)
    {
        if (!Directory.Exists(root))
        {
            return Task.CompletedTask;
        }

        return Task.Run(() => Directory.Delete(root, recursive: true));
    }
}


